packages/compartment-mapper/src/types/internal.d.ts

```diff
@@ -0,0 +1,218 @@
+/**
+ * @file Internal types of the compartment mapper that need not be visible to
+ * consumers.
+ */
+
+/* eslint-disable no-use-before-define */
+
+import type { ImportHook, ImportNowHook } from 'ses';
+import type {
+  CompartmentDescriptor,
+  Language,
+  LanguageForExtension,
+  LanguageForModuleSpecifier,
+  ModuleDescriptor,
+} from './compartment-map-schema.ts';
+import type {
+  HashFn,
+  MaybeReadFn,
+  MaybeReadNowFn,
+  ReadFn,
+  ReadPowers,
+} from './powers.ts';
+import type { DeferredAttenuatorsProvider } from './policy.ts';
+import type {
+  AsyncParseFn,
+  CompartmentSources,
+  ExecuteOptions,
+  ExitModuleImportNowHook,
+  LanguageForExtensionOption,
+  ModuleTransforms,
+  ParseFn,
+  ParserForLanguage,
+  SearchSuffixesOption,
+  SourceMapHook,
+  SourceMapHookOption,
+  Sources,
+  SyncModuleTransforms,
+} from './external.ts';
+
+export type LinkOptions = {
+  resolve?: ResolveHook;
+  makeImportHook: ImportHookMaker;
+  makeImportNowHook?: ImportNowHookMaker;
+  parserForLanguage?: ParserForLanguage;
+  moduleTransforms?: ModuleTransforms;
+  syncModuleTransforms?: SyncModuleTransforms;
+  archiveOnly?: boolean;
+} & ExecuteOptions &
+  LanguageForExtensionOption;
+
+export type LinkResult = {
+  compartment: Compartment;
+  compartments: Record<string, Compartment>;
+  attenuatorsCompartment: Compartment;
+  pendingJobsPromise: Promise<void>;
+};
+
+export type ResolveHook = (
+  importSpecifier: string,
+  referrerSpecifier: string,
+) => string;
+
+export type ShouldDeferError = (language: Language | undefined) => boolean;
+
+export type MakeImportNowHookMakerOptions = Partial<{
+  sources: Sources;
+  compartmentDescriptors: Record<string, CompartmentDescriptor>;
+  computeSha512: HashFn;
+  /**
+   */
+  exitModuleImportNowHook: ExitModuleImportNowHook;
+}> &
+  SearchSuffixesOption &
+  SourceMapHookOption;
+
+export type ImportHookMaker = (params: {
+  packageLocation: string;
+  packageName: string;
+  attenuators: DeferredAttenuatorsProvider;
+  parse: ParseFn | AsyncParseFn;
+  shouldDeferError: ShouldDeferError;
+  compartments: Record<string, Compartment>;
+}) => ImportHook;
+
+export type ImportNowHookMaker = (params: {
+  packageLocation: string;
+  packageName: string;
+  parse: ParseFn | AsyncParseFn;
+  compartments: Record<string, Compartment>;
+  // TODO why are these absent? Why this asymmetry?
+  // attenuators: DeferredAttenuatorsProvider,
+  // shouldDeferError: ShouldDeferError,
```

Comment:
they are ignored/unused

Reply:
...and maybe we should consider adding `attenuators: never, shouldDeferError: never` to this type, just so whoever is using it cannot expect any behavior based on these two props

Reply:
I’m patching this as advised, and documenting the deliberate discrepancy, but is there a principled argument why they are not respected, to the extent that it would not make sense to add behavior for these properties as a future feature, or are they omitted for lack of a motivating use case?

Reply:
I _thought_ I understood why `shouldDeferError` was omitted, but now I'm not so sure.  I _think_ it's essentially "too late" to defer an error.  @naugtur, can you explain this (again)?  

I'm assuming `DeferredAttenuatorsProvider` is not provided because a) there's only one attenuator compartment and b) that compartment is loaded via `importHook` instead of `importNowHook`.